dart/dynamics/IkFast.cpp

```diff
@@ -0,0 +1,329 @@
+      break;
+    }
+
+    index++;
+  }
+
+  solution.mValidity
+      = limitViolated ? InverseKinematics::Analytical::LIMIT_VIOLATED
+                      : InverseKinematics::Analytical::VALID;
+}
+
+//==============================================================================
+void convertIkSolutions(
+    const IkFast* ikfast,
+    int numJoints,
+    int numFreeParameters,
+    int* freeParameters,
+    const ikfast::IkSolutionList<IkReal>& ikfastSolutions,
+    std::vector<InverseKinematics::Analytical::Solution>& solutions)
+{
+  const auto numSolutions = ikfastSolutions.GetNumSolutions();
+
+  solutions.resize(numSolutions);
+
+  for (auto i = 0u; i < numSolutions; ++i)
+  {
+    const auto& ikfastSolution = ikfastSolutions.GetSolution(i);
+    convertIkSolution(
+        ikfast,
+        numJoints,
+        numFreeParameters,
+        freeParameters,
+        ikfastSolution,
+        solutions[i]);
+  }
+}
+
+} // namespace (anonymous)
+
+//==============================================================================
+IkFast::IkFast(
+    InverseKinematics* ik,
+    const std::vector<std::size_t>& dofMap,
+    const std::vector<std::size_t>& freeDofMap,
+    const std::string& methodName,
+    const InverseKinematics::Analytical::Properties& properties)
+  : Analytical{ik, methodName, properties}, mConfigured{false}
+{
+  setExtraDofUtilization(UNUSED);
+
+  mDofs = dofMap;
+  mFreeDofs = freeDofMap;
+}
+
+//==============================================================================
+auto IkFast::getDofs() const -> const std::vector<std::size_t>&
+{
+  if (!mConfigured)
+  {
+    configure();
+
+    if (!mConfigured)
+    {
+      dtwarn << "[IkFast::getDofs] This analytical IK was not able "
+             << "to configure properly, so it will not be able to compute "
+             << "solutions. Returning an empty list of dofs.\n";
+      assert(mDofs.empty());
+    }
+  }
+
+  return mDofs;
+}
+
+//==============================================================================
+bool IkFast::isConfigured() const
+{
+  return mConfigured;
+}
+
+//==============================================================================
+bool checkDofMapValidity(
+    const InverseKinematics* ik,
+    const std::vector<std::size_t>& dofMap,
+    const std::vector<DegreeOfFreedom*>& dependentDofs,
+    const std::string& dofMapName)
+{
+  for (const auto& dof : dofMap)
+  {
+    bool found = false;
+    for (auto dependentDof : dependentDofs)
+    {
+      const auto dependentDofIndex = dependentDof->getIndexInSkeleton();
+      if (dof == dependentDofIndex)
+      {
+        found = true;
+        break;
+      }
+    }
+
+    if (!found)
+    {
+      dterr << "[IkFast::configure] Failed to configure. An element of the "
+            << "given " << dofMapName << " '" << dof
+            << "' is not a dependent dofs of Node '" << ik->getNode()->getName()
+            << "'.\n";
+      return false;
+    }
+  }
+
+  return true;
+}
```

Comment:
I think this function is trying to verify that all of the DOFs effect the `EndEffector` frame. However, I am having ar really hard time following the implementation. Can you add a few comments?